pennylane/beta/tapes/qnode.py

```diff
@@ -97,6 +98,10 @@ class QNode:
             * ``"finite-diff"``: Uses numerical finite-differences for all quantum operation
               arguments.
 
+        caching (int): number of device executions to store in a cache to speed up subsequent
+            executions. Caching does not take place by default. In caching mode, the quantum circuit
+            being executed must have a constant structure and only its parameters can be varied.

```

Comment:
In addition (not sure if it's implied by 'only its parameters can be varied', but it wasn't clear to me): there can't be any classical processing within the QNode.

E.g., `qml.RX(2*x, wires=0)` is not allowed.

This is a side effect of the tape being a part of the users quantum function --- we can't extract the tape independently of the classical processing :(

Reply:
I have to admit, it's not something I've tested, just a 'hunch' that it wouldn't work. My thinking was that:

When the QNode is evaluated, the users `qfunc()` is called _inside_ a quantum tape:

```python
with QuantumTape() as tape:
    qfunc(*args, **kwargs)
```

This will do two things:

1. classically process the `args` and `kwargs` as per whatever classical processing the user has within the QNode, e.g., `tf.sin(a)`

2. construct the tape, with incident gate arguments potentially different to `args` and `kwargs` due to post-processing.

Say the user passes a qfunc of the form:

```python
def circuit(x):
    qml.RX(tf.sin(x), wires=0)
    return qml.expval(qml.PauliZ(0))
```

After the first construction (the user has passed x=0.6), we 'cache' or store the tape. On the second evaluation (now the user passes x=0.1), we want to execute the _previous_ tape, but with the new gate argument; however, we can't do `tape.set_parameters([0.1])`, because the user actually wants `tape.set_parameters([tf.sin(0.1)])` --- how do we determine the classical post-processing step?

So having immutable tapes is not enough as far as I can see. Or maybe I am missing something --- would be happy if that's the case and this works!